README.md
# ScrollGalleryView
[![](https://jitpack.io/v/javaherisaber/ScrollGalleryView.svg)](https://jitpack.io/#javaherisaber/ScrollGalleryView)
[![Android Arsenal](https://img.shields.io/badge/Android%20Arsenal-ScrollGalleryView-brightgreen.svg?style=flat)](https://android-arsenal.com/details/1/2472)

ScrollGalleryView is a flexible library which helps you to create awesome media galleries in your Android application. It's easily integrated with the most popular image loading libraries such as Picasso, Glide and Fresco.

![ScrollGalleryView](http://i.imgur.com/xrBt4Xx.gif)

## Key features

- Easy way to select images in gallery (thumbnails)
- Zooming
- Simple API
- Video

## Installing

Add [JitPack](https://jitpack.io) repository to your root `build.gradle`:

```gradle
allprojects {
    repositories {
        // ...
        maven { url 'https://jitpack.io' }
    }
}
```

Add dependency to your module `build.gradle`:

```gradle
implementation "com.github.javaherisaber:ScrollGalleryView:$versions.scrollGalleryView" // all libs (including media loaders)
implementation "com.github.javaherisaber.ScrollGalleryView:library:$versions.scrollingGalleryView" // core library only
```

### MediaLoaders

There are several MediaLoaders implementations for most popular caching libraries: Picasso, Glide, Fresco.

#### Picasso

```gradle
implementation "com.github.javaherisaber.ScrollGalleryView:picasso-loader:$versions.scrollingGalleryView"
```

#### Glide

```gradle
implementation "com.github.javaherisaber.ScrollGalleryView:glide-loader:$versions.scrollingGalleryView"
```

#### Fresco

```gradle
implementation "com.github.javaherisaber.ScrollGalleryView:fresco-loader:$versions.scrollingGalleryView"
```

## Usage

Add *ScrollGalleryView* to your layout:

```xml
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
    android:layout_width="match_parent"
    android:layout_height="match_parent">
    <com.veinhorn.scrollgalleryview.ScrollGalleryView
        android:id="@+id/scroll_gallery_view"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:background="#000"/>
</LinearLayout>
```

Initialize *ScrollGalleryView* in your activity:

```java
import static com.veinhorn.scrollgalleryview.loader.picasso.DSL.*; // simplifies adding media

public class MainActivity extends FragmentActivity {
    private ScrollGalleryView galleryView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ScrollGalleryView galleryView = findViewById(R.id.scroll_gallery_view);
        galleryView.setFragmentManager(getSupportFragmentManager())
                .setThumbnailSize(150)
                .setZoom(true)
                .withHiddenThumbnails(false)
                .hideThumbnailsOnClick(true)
                .setSelectedIndex(2)
                .addMedia(image("https://i.picsum.photos/id/249/960/1080.jpg?hmac=M1eYv2hPaxVrc1jgoI4o9r5YJi_Xm9kMEbaxFfQn-yU"))
                .addMedia(image("https://i.picsum.photos/id/803/960/1080.jpg?hmac=HFNjRwjV8mL2POgQh45zzTc1MkBU55vgxbpwJeg7utk"))
                .addMedia(image("https://i.picsum.photos/id/976/960/1080.jpg?hmac=Osd25GIpQNNUZe7y__Leq_FL83JmeoiOzVdn2CrE7DM"))
                .addMedia(image("https://i.picsum.photos/id/593/960/1080.jpg?hmac=vFbBrBQQUKOhqDcf_1fMvKs6KzlF21qqS04X2tzbumo"))
                .addMedia(image("https://i.picsum.photos/id/932/960/1080.jpg?hmac=DTJwQ2y6OSfcuFI6mY8rpmtWjliCxqzP1-NsVC-nICM"))
                .build();
    }
}
```

If you use *ScrollGalleryView* version prior `1.2.0` or need more info about gallery initialization you can find it [here](docs/init-gallery.md).

### Adding media

ScrollGalleryView supports different types of media such as images and videos. You can create image gallery, video gallery, or mix them in any way. To abstract from concreate way of image loading ScrollGalleryView uses [MediaLoader](library/src/main/java/com/veinhorn/scrollgalleryview/loader/MediaLoader.java) so it makes possible to use different image loading libraries depending on your needs (Picasso, Glide, Fresco).

#### **Picasso** loader

#### Default image loader

Library is also contains default image loader but it's not optimized for performance.

> Note: it's highly recommended to use custom image loader against default

### Configuration

You can specify a bunch of additional settings during gallery initialization.

|Option|Method|Description|
|------|------|-----------|
| Thumbnail size | `.setThumbnailSize(200)` | You can configure thumbnails size in |
| Zoom | `.setZoom(true)` | Enable zoom |
| Hide thumbnails | `.withHiddenThumbnails(false)` | Hide scroll view container with thumbnails on the bottom of screen |
| Hide thumbnails on click | `.hideThumbnailsOnClick(true)` | Hide scroll view container with thumbnails when you click on main image area |

### Adding listeners

*ScrollGalleryView* supports adding listeners for events like: image click, long image click, changed page. More details you can find in [separate doc](docs/event-listeners.md).

## Sample application

The sample application published on Google Play.

[![Get it on Google Play](http://www.android.com/images/brand/get_it_on_play_logo_small.png)](https://play.google.com/store/apps/details?id=com.veinhorn.scrollgalleryview)

## License

    MIT License
    
    Copyright (c) 2019 Boris Korogvich
    
    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.
    
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.


CHANGELOG.md
# 1.3.1
- Add onPageSelected listener
- Upgrade AGP to 4.1.1

# 1.3.0
- Fix crash when using Glide 4.11.0
- Flatten versioning for all artifacts
- Force ltr direction for ViewPager and HorizontalScrollView
- Add close button
- Add position text
- Add selected thumbnail indicator
- Show controls when hided and user scroll to next position
- Add selected item and scroll to position first time layout is drawn

library/src/main/java/com/veinhorn/scrollgalleryview/ScrollGalleryView.java
package com.veinhorn.scrollgalleryview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Point;
import android.graphics.drawable.BitmapDrawable;
import android.media.ThumbnailUtils;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Pair;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.HorizontalScrollView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import androidx.fragment.app.FragmentManager;
import androidx.transition.Transition;
import androidx.transition.TransitionManager;
import androidx.viewpager.widget.ViewPager;

import com.veinhorn.scrollgalleryview.builder.GalleryBuilder;
import com.veinhorn.scrollgalleryview.builder.GalleryBuilderImpl;
import com.veinhorn.scrollgalleryview.loader.MediaLoader;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ScrollGalleryView extends LinearLayout {
    private FragmentManager fragmentManager;
    private Context context;
    private Point displayProps;
    private ScreenSlidePagerAdapter pagerAdapter;
    private List<MediaInfo> mListOfMedia;
    private int selectedIndex = 0;

    // Options
    private int thumbnailSize; // width and height in pixels
    private boolean zoomEnabled;
    private boolean isControlsHidden;
    private boolean hideControlsOnClick;
    private Integer hideControlsAfterDelay;

    // Views
    private LinearLayout thumbnailsContainer;
    private HorizontalScrollView horizontalScrollView;
    private ViewPager viewPager;
    public TextView positionView;
    public ImageView closeView;

    // Transitions
    private Transition thumbnailsTransition;
    private boolean useDefaultThumbnailsTransition;

    public ScrollGalleryView(Context context) {
        super(context);
        init(context, null);
    }

    public ScrollGalleryView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public ScrollGalleryView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public ScrollGalleryView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ScrollGalleryView);
        // Attrs
        int closeIconColor = array.getColor(R.styleable.ScrollGalleryView_closeIconColor, Color.WHITE);
        int countTextColor = array.getColor(R.styleable.ScrollGalleryView_countTextColor, Color.WHITE);
        int countTextSize = array.getDimensionPixelSize(R.styleable.ScrollGalleryView_countTextSize, -1);
        array.recycle();

        this.context = context;
        mListOfMedia = new ArrayList<>();

        setOrientation(VERTICAL);
        displayProps = getDisplaySize();
        LayoutInflater inflater = LayoutInflater.from(context);
        inflater.inflate(R.layout.scroll_gallery_view, this, true);

        closeView = (ImageView) findViewById(R.id.imageView_close);
        closeView.setColorFilter(closeIconColor);

        positionView = (TextView) findViewById(R.id.textView_position);
        positionView.setTextColor(countTextColor);
        if (countTextSize != -1) {
            positionView.setTextSize(pixelToDp(context, countTextSize));
        }

        horizontalScrollView = (HorizontalScrollView) findViewById(R.id.thumbnails_scroll_view);
        thumbnailsContainer = (LinearLayout) findViewById(R.id.thumbnails_container);
        thumbnailsContainer.setPadding(displayProps.x / 2, 0, displayProps.x / 2, 0);
    }

    private int pixelToDp(Context context, int pixel) {
        return (int) (pixel / context.getResources().getDisplayMetrics().density);
    }

    public interface OnPageSelectedListener {
        void onPageSelected(int position);
    }

    private OnPageSelectedListener onPageSelectedListener;

    public void setOnPageSelectedListener(OnPageSelectedListener listener) {
        this.onPageSelectedListener = listener;
    }

    // Listeners
    private final ViewPager.SimpleOnPageChangeListener viewPagerChangeListener = new ViewPager.SimpleOnPageChangeListener() {
        @Override
        public void onPageSelected(int position) {
            updateSelectedThumbnail(false);
            selectedIndex = position;
            updateSelectedThumbnail(true);
            showControls();
            updatePositionView();
            scroll(thumbnailsContainer.getChildAt(selectedIndex));
            if (onPageSelectedListener != null) {
                onPageSelectedListener.onPageSelected(position);
            }
        }
    };

    private void updatePositionView() {
        positionView.setText((selectedIndex + 1) + "/" + mListOfMedia.size());
    }

    private void updateSelectedThumbnail(boolean isActive) {
        View view = thumbnailsContainer.getChildAt(selectedIndex);
        View activeView = view.findViewById(R.id.view_thumbnail_active);
        if (isActive) {
            activeView.setVisibility(View.VISIBLE);
        } else  {
            activeView.setVisibility(View.INVISIBLE);
        }
    }

    private final OnClickListener thumbnailOnClickListener = new OnClickListener() {
        @Override
        public void onClick(View v) {
            int position = (int) v.getTag();
            updateSelectedThumbnail(false);
            selectedIndex = position;
            updateSelectedThumbnail(true);
            scroll(v);
            viewPager.setCurrentItem(position, true);
        }
    };

    private OnImageClickListener onImageClickListener;
    private OnImageLongClickListener onImageLongClickListener;

    /**
     * We should create OnImageClickListener to wrap our provided OnImageClickListener,
     * otherwise it will be null when we pass it to PagerAdapter. Also it used to wrap
     * code which is responsible for showing/hiding thumbnails when user click on image
     */
    private OnImageClickListener innerOnImageClickListener = new OnImageClickListener() {
        @Override
        public void onClick(int position) {
            if (hideControlsOnClick) {
                if (isControlsHidden) {
                    showControls();
                    isControlsHidden = false;
                } else {
                    hideControls();
                    isControlsHidden = true;
                }
            }
            if (onImageClickListener != null) onImageClickListener.onClick(position);
        }
    };

    private OnImageLongClickListener innerOnImageLongClickListener = new OnImageLongClickListener() {
        @Override
        public void onClick(int position) {
            if (onImageLongClickListener != null) onImageLongClickListener.onClick(position);
        }
    };

    public interface OnImageClickListener {
        void onClick(int position);
    }

    public interface OnImageLongClickListener {
        void onClick(int position);
    }

    public ScrollGalleryView setFragmentManager(FragmentManager fragmentManager) {
        this.fragmentManager = fragmentManager;
        initializeViewPager();

        if (hideControlsAfterDelay != null) hideThumbnailsAfterDelay(hideControlsAfterDelay);

        return this;
    }

    public void build() {
        horizontalScrollView.post(new Runnable() {
            @Override
            public void run() {
                scroll(thumbnailsContainer.getChildAt(selectedIndex));
            }
        });
    }

    /**
     * @return inner ViewPager
     */
    public ViewPager getViewPager() {
        return viewPager;
    }

    /**
     * Set up OnImageClickListener for your gallery images
     * You should set OnImageClickListener only before setFragmentManager call!
     *
     * @param onImageClickListener which is called when you click on image
     * @return ScrollGalleryView
     */
    public ScrollGalleryView addOnImageClickListener(OnImageClickListener onImageClickListener) {
        this.onImageClickListener = onImageClickListener;
        return this;
    }

    public ScrollGalleryView addOnImageLongClickListener(OnImageLongClickListener onImageLongClickListener) {
        this.onImageLongClickListener = onImageLongClickListener;
        return this;
    }

    /**
     * Set up OnPageChangeListener for internal ViewPager
     *
     * @param listener which is used by internal ViewPager
     */
    public void addOnPageChangeListener(final ViewPager.OnPageChangeListener listener) {
        viewPager.clearOnPageChangeListeners();
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                listener.onPageScrolled(position, positionOffset, positionOffsetPixels);
            }

            @Override
            public void onPageSelected(int position) {
                scroll(thumbnailsContainer.getChildAt(position));
                listener.onPageSelected(position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                listener.onPageScrollStateChanged(state);
            }
        });
    }

    public ScrollGalleryView addMedia(MediaInfo mediaInfo) {
        if (mediaInfo == null) {
            throw new NullPointerException("Infos may not be null!");
        }

        return addMedia(Collections.singletonList(mediaInfo));
    }

    public ScrollGalleryView setSelectedIndex(int index) {
        this.selectedIndex = index;
        return this;
    }

    public ScrollGalleryView addMedia(List<MediaInfo> infos) {
        if (infos == null) {
            throw new NullPointerException("Infos may not be null!");
        }

        for (int index = 0; index < infos.size(); index++) {
            MediaInfo info = infos.get(index);
            mListOfMedia.add(info);

            final ImageView thumbnail = addThumbnail(getDefaultThumbnail(), selectedIndex == mListOfMedia.size() - 1);
            info.getLoader().loadThumbnail(getContext(), thumbnail, new MediaLoader.SuccessCallback() {
                @Override
                public void onSuccess() {
                    thumbnail.setScaleType(ImageView.ScaleType.FIT_CENTER);
                }
            });

            pagerAdapter.notifyDataSetChanged();
        }
        if (selectedIndex < mListOfMedia.size()) {
            viewPager.setCurrentItem(selectedIndex, true);
        }
        updatePositionView();

        return this;
    }

    private ImageView addThumbnail(Bitmap image, boolean isSelected) {
        Bitmap thumbnail = createThumbnail(image);
        Pair<View, ImageView> thumbnailView = createThumbnailView(thumbnail, isSelected);
        thumbnailsContainer.addView(thumbnailView.first);
        if (thumbnailsContainer.getChildCount() > selectedIndex) {
            scroll(thumbnailsContainer.getChildAt(selectedIndex));
        }
        return thumbnailView.second;
    }

    private Pair<View, ImageView> createThumbnailView(Bitmap thumbnail, boolean isSelected) {
        View view = inflate(context, R.layout.thumbnail_view, null);
        ImageView imageView = (ImageView) view.findViewById(R.id.imageView_thumbnail);
        View active = (View) view.findViewById(R.id.view_thumbnail_active);
        imageView.setImageBitmap(thumbnail);
        imageView.setOnClickListener(thumbnailOnClickListener);
        imageView.setTag(mListOfMedia.size() - 1);
        imageView.setScaleType(ImageView.ScaleType.CENTER);
        if (isSelected) {
            active.setVisibility(View.VISIBLE);
        } else {
            active.setVisibility(View.GONE);
        }
        return new Pair<>(view, imageView);
    }

    /**
     * Set the current item displayed in the view pager.
     *
     * @param i a zero-based index
     */
    public ScrollGalleryView setCurrentItem(int i) {
        viewPager.setCurrentItem(i, false);
        return this;
    }

    public int getCurrentItem() {
        return viewPager.getCurrentItem();
    }

    public ScrollGalleryView setThumbnailSize(int thumbnailSize) {
        this.thumbnailSize = thumbnailSize;
        return this;
    }

    public ScrollGalleryView setZoom(boolean zoomEnabled) {
        this.zoomEnabled = zoomEnabled;
        return this;
    }

    /**
     * If you enabled this option, hideThumbnailsOnClick() method will not work
     *
     * @param isThumbnailsHidden hides thumbnails container
     * @return ScrollGalleryView
     */
    public ScrollGalleryView withHiddenThumbnails(boolean isThumbnailsHidden) {
        if (this.isControlsHidden && !isThumbnailsHidden) {
            showControls();
        } else if (!this.isControlsHidden && isThumbnailsHidden) {
            hideControls();
        }
        this.isControlsHidden = isThumbnailsHidden;

        return this;
    }

    /**
     * Keep in mind that this method do not work with enabled isThumbnailsHidden option
     *
     * @param hideThumbnailsOnClick hides thumbnails container on image click
     * @return ScrollGalleryView
     */
    public ScrollGalleryView hideThumbnailsOnClick(boolean hideThumbnailsOnClick) {
        if (!isControlsHidden) {
            this.hideControlsOnClick = hideThumbnailsOnClick;
            if (hideThumbnailsOnClick) this.useDefaultThumbnailsTransition = true;
        }
        return this;
    }

    /**
     * Keep in mind that this method do not work with enabled isThumbnailsHidden option
     *
     * @param hideThumbnailsOnClick hides thumbnails container on image click
     * @param thumbnailsTransition  null is used to disable transation
     * @return ScrollGalleryView
     */
    public ScrollGalleryView hideThumbnailsOnClick(boolean hideThumbnailsOnClick, Transition thumbnailsTransition) {
        if (!isControlsHidden) {
            this.hideControlsOnClick = hideThumbnailsOnClick;
            this.thumbnailsTransition = thumbnailsTransition;
        }
        return this;
    }

    /**
     * Automatically hide thumbnails container after specified delay
     *
     * @param hideThumbnailsAfterDelay delay in ms
     * @return ScrollGalleryView
     */
    public ScrollGalleryView hideThumbnailsAfter(int hideThumbnailsAfterDelay) {
        if (!isControlsHidden) {
            this.hideControlsAfterDelay = hideThumbnailsAfterDelay;
        }
        return this;
    }

    public void showControls() {
        closeView.setVisibility(View.VISIBLE);
        positionView.setVisibility(View.VISIBLE);
        setThumbnailsTransition();
        horizontalScrollView.setVisibility(VISIBLE);
        // Hide thumbnails container when hideThumbnailsAfterDelay option is enabled
        if (hideControlsAfterDelay != null) hideThumbnailsAfterDelay(hideControlsAfterDelay);
    }

    public void hideControls() {
        closeView.setVisibility(View.GONE);
        positionView.setVisibility(View.GONE);
        setThumbnailsTransition();
        horizontalScrollView.setVisibility(GONE);
    }

    /**
     * Remove all images from gallery
     */
    public void clearGallery() {
        // remove all media infos
        mListOfMedia.clear();

        // create new adapter
        pagerAdapter = new ScreenSlidePagerAdapter(fragmentManager, mListOfMedia, zoomEnabled, innerOnImageClickListener,
                innerOnImageLongClickListener,
                getResources().getString(R.string.first_image_transition_name));

        viewPager.setAdapter(pagerAdapter);
        // remove thumbnails
        thumbnailsContainer.removeAllViews();
    }

    /**
     * Remove a media from the gallery
     *
     * @param position media's position to remove
     */
    public void removeMedia(int position) {
        if (position >= mListOfMedia.size() || position < 0) {
            return;
        }
        pagerAdapter.removeItem(position);
        removeThumbnail(position);
    }

    public static GalleryBuilder from(ScrollGalleryView galleryView) {
        return new GalleryBuilderImpl(galleryView);
    }

    private void hideThumbnailsAfterDelay(int delay) {
        horizontalScrollView.postDelayed(new Runnable() {
            @Override
            public void run() {
                hideControls();
                isControlsHidden = !isControlsHidden;
            }
        }, delay);
    }

    // Make choice between default and provided by user transition
    private void setThumbnailsTransition() {
        if (thumbnailsTransition == null && useDefaultThumbnailsTransition) {
            TransitionManager.beginDelayedTransition(horizontalScrollView);
        } else if (thumbnailsTransition != null) {
            TransitionManager.beginDelayedTransition(horizontalScrollView, thumbnailsTransition);
        }
    }

    private Bitmap getDefaultThumbnail() {
        return ((BitmapDrawable) getContext().getResources().getDrawable(R.drawable.placeholder_image)).getBitmap();
    }

    private Point getDisplaySize() {
        WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        Point point = new Point();
        display.getSize(point);
        return point;
    }

    private void removeThumbnail(int position) {
        View thumbnail = thumbnailsContainer.getChildAt(position);
        if (thumbnail == null) {
            return;
        }
        thumbnailsContainer.removeView(thumbnail);
    }

    private Bitmap createThumbnail(Bitmap image) {
        return ThumbnailUtils.extractThumbnail(image, thumbnailSize, thumbnailSize);
    }

    private void initializeViewPager() {
        viewPager = (HackyViewPager) findViewById(R.id.viewPager);

        pagerAdapter = new ScreenSlidePagerAdapter(
                fragmentManager,
                mListOfMedia,
                zoomEnabled,
                innerOnImageClickListener,
                innerOnImageLongClickListener,
                getResources().getString(R.string.first_image_transition_name));

        viewPager.setAdapter(pagerAdapter);
        viewPager.addOnPageChangeListener(viewPagerChangeListener);
    }

    private void scroll(View thumbnail) {
        int thumbnailCoords[] = new int[2];
        thumbnail.getLocationOnScreen(thumbnailCoords);

        int thumbnailCenterX = thumbnailCoords[0] + thumbnailSize / 2;
        int thumbnailDelta = displayProps.x / 2 - thumbnailCenterX;

        horizontalScrollView.smoothScrollBy(-thumbnailDelta, 0);
    }

    private int calculateInSampleSize(int imgWidth, int imgHeight, int maxWidth, int maxHeight) {
        int inSampleSize = 1;
        while (imgWidth / inSampleSize > maxWidth || imgHeight / inSampleSize > maxHeight) {
            inSampleSize *= 2;
        }
        return inSampleSize;
    }
}


glide-loader/src/main/java/ogbe/ozioma/com/glideimageloader/GlideImageLoader.java
package ogbe.ozioma.com.glideimageloader;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.veinhorn.scrollgalleryview.loader.MediaLoader;

public class GlideImageLoader implements MediaLoader {

    private String url;
    private RequestOptions requestOptions;

    public GlideImageLoader(String url) {
        this.url = url;
        requestOptions = new RequestOptions()
                .placeholder(R.drawable.placeholder_image);
    }

    @SuppressWarnings("unused")
    public GlideImageLoader(String url, int width, int height) {
        this.url = url;
        requestOptions = new RequestOptions()
                .placeholder(R.drawable.placeholder_image)
                .override(width, height);
    }

    @Override
    public boolean isImage() {
        return true;
    }

    @Override
    public void loadMedia(Context context, ImageView imageView, final SuccessCallback callback) {
        RequestOptions requestOptions = new RequestOptions();
        Glide.with(context).applyDefaultRequestOptions(requestOptions)
                .load(url).listener(new RequestListener<Drawable>() {
            @Override
            public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                return false;
            }

            @Override
            public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                callback.onSuccess();
                return false;
            }
        }).into(imageView);
    }

    @Override
    public void loadThumbnail(Context context, ImageView thumbnailView, final SuccessCallback callback) {
        Glide.with(context).applyDefaultRequestOptions(requestOptions).load(url).listener(new RequestListener<Drawable>() {
            @Override
            public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                return false;
            }

            @Override
            public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                callback.onSuccess();
                return false;
            }
        }).into(thumbnailView);
    }

}


library/src/main/java/com/veinhorn/scrollgalleryview/ScreenSlidePagerAdapter.java
package com.veinhorn.scrollgalleryview;

import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.viewpager.widget.PagerAdapter;

import java.util.List;

/**
 * Created by veinhorn on 29.8.15.
 */
public class ScreenSlidePagerAdapter extends FragmentStatePagerAdapter {

    private List<MediaInfo> mListOfMedia;

    private boolean isZoom = false;

    private ScrollGalleryView.OnImageClickListener onImageClickListener;
    private ScrollGalleryView.OnImageLongClickListener onImageLongClickListener;
    
    private String transitionName;

    public ScreenSlidePagerAdapter(FragmentManager fm, List<MediaInfo> listOfMedia,
                                   boolean isZoom, ScrollGalleryView.OnImageClickListener onImageClickListener,
                                   ScrollGalleryView.OnImageLongClickListener onImageLongClickListener,
                                   String transitionName) {
        super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        this.mListOfMedia = listOfMedia;
        this.isZoom = isZoom;
        this.onImageClickListener = onImageClickListener;
        this.transitionName = transitionName;

        this.onImageLongClickListener = onImageLongClickListener;
    }

    @Override
    public Fragment getItem(int position) {
        Fragment fragment = null;
        if (position < mListOfMedia.size()) {
            MediaInfo mediaInfo = mListOfMedia.get(position);
            fragment = loadImageFragment(mediaInfo, position);
            if (position == 0)
                ((ImageFragment)fragment).setTransitionName(transitionName);
        }
        return fragment;
    }

    private Fragment loadImageFragment(MediaInfo mediaInfo, int position) {
        ImageFragment fragment = new ImageFragment();
        fragment.setRetainInstance(true);
        fragment.setMediaInfo(mediaInfo);

        if (onImageClickListener != null) {
            fragment.setOnImageClickListener(onImageClickListener);
        }
        if (onImageLongClickListener != null) {
            fragment.setOnImageLongClickListener(onImageLongClickListener);
        }

        Bundle bundle = new Bundle();
        bundle.putBoolean(Constants.ZOOM, isZoom);
        bundle.putInt("position", position);

        fragment.setArguments(bundle);
        return fragment;
    }

    public void removeItem(int position) {
        mListOfMedia.remove(position);
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return mListOfMedia.size();
    }

    @Override
    public int getItemPosition(Object object) {
        return PagerAdapter.POSITION_NONE;
    }
}


picasso-loader/src/main/java/com/veinhorn/scrollgalleryview/loader/picasso/PicassoImageLoader.java
package com.veinhorn.scrollgalleryview.loader.picasso;

import android.content.Context;
import android.widget.ImageView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;
import com.veinhorn.scrollgalleryview.loader.MediaLoader;

/**
 * Created by veinhorn on 2/4/18.
 */
public class PicassoImageLoader implements MediaLoader {
    private String url;
    private Integer thumbnailWidth;
    private Integer thumbnailHeight;

    public PicassoImageLoader(String url) {
        this.url = url;
    }

    public PicassoImageLoader(String url, Integer thumbnailWidth, Integer thumbnailHeight) {
        this.url = url;
        this.thumbnailWidth = thumbnailWidth;
        this.thumbnailHeight = thumbnailHeight;
    }

    @Override
    public boolean isImage() {
        return true;
    }

    @Override
    public void loadMedia(Context context, final ImageView imageView, final MediaLoader.SuccessCallback callback) {
        Picasso.get()
                .load(url)
                .placeholder(R.drawable.placeholder_image)
                .into(imageView, new ImageCallback(callback));
    }

    @Override
    public void loadThumbnail(Context context, final ImageView thumbnailView, final MediaLoader.SuccessCallback callback) {
        Picasso.get()
                .load(url)
                .resize(thumbnailWidth == null ? 100 : thumbnailWidth,
                        thumbnailHeight == null ? 100 : thumbnailHeight)
                .placeholder(R.drawable.placeholder_image)
                .centerInside()
                .into(thumbnailView, new ImageCallback(callback));
    }

    private static class ImageCallback implements Callback {
        private final MediaLoader.SuccessCallback callback;

        public ImageCallback(SuccessCallback callback) {
            this.callback = callback;
        }

        @Override public void onSuccess() {
            callback.onSuccess();
        }

        @Override public void onError(Exception e) {
        }
    }
}


app/src/main/java/com/veinhorn/example/MainActivity.java
package com.veinhorn.example;

import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import com.veinhorn.scrollgalleryview.MediaInfo;
import com.veinhorn.scrollgalleryview.ScrollGalleryView;
import ogbe.ozioma.com.glideimageloader.GlideImageLoader;

import static ogbe.ozioma.com.glideimageloader.DSL.image;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ScrollGalleryView scrollGalleryView = findViewById(R.id.scroll_gallery_view);
        scrollGalleryView.setFragmentManager(getSupportFragmentManager())
                .setThumbnailSize(150)
                .setZoom(true)
                .withHiddenThumbnails(false)
                .hideThumbnailsOnClick(true)
                .setSelectedIndex(2)
                .addMedia(image("https://i.picsum.photos/id/249/960/1080.jpg?hmac=M1eYv2hPaxVrc1jgoI4o9r5YJi_Xm9kMEbaxFfQn-yU"))
                .addMedia(image("https://i.picsum.photos/id/803/960/1080.jpg?hmac=HFNjRwjV8mL2POgQh45zzTc1MkBU55vgxbpwJeg7utk"))
                .addMedia(image("https://i.picsum.photos/id/976/960/1080.jpg?hmac=Osd25GIpQNNUZe7y__Leq_FL83JmeoiOzVdn2CrE7DM"))
                .addMedia(image("https://i.picsum.photos/id/593/960/1080.jpg?hmac=vFbBrBQQUKOhqDcf_1fMvKs6KzlF21qqS04X2tzbumo"))
                .addMedia(image("https://i.picsum.photos/id/932/960/1080.jpg?hmac=DTJwQ2y6OSfcuFI6mY8rpmtWjliCxqzP1-NsVC-nICM"))
                .build();

        scrollGalleryView.closeView.setOnClickListener(view -> finish());
    }
}
